source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -1236,6 +1236,56 @@ bool COpenGLDriver::beginScene(bool backBuffer, bool zBuffer, SColor color,
 }
 
 
+bool COpenGLDriver::bindGraphicsPipeline(video::IGPURenderpassIndependentPipeline* _gpipeline)
+{
+    SAuxContext* ctx = getThreadContext_helper(false);
+    if (!ctx)
+        return false;
+
+    COpenGLRenderpassIndependentPipeline* glpipeline = static_cast<COpenGLRenderpassIndependentPipeline*>(_gpipeline);
+    ctx->nextState.pipeline = core::smart_refctd_ptr<COpenGLRenderpassIndependentPipeline>(glpipeline);
+
+    //all code below this line is descriptor set bindings invalidation according to pipeline layout compatibility rules (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#descriptorsets-compatibility)
+    if (!ctx->currentState.pipeline || !_gpipeline)
+    {
+        std::fill(ctx->nextState.descriptorsParams.descSets, ctx->nextState.descriptorsParams.descSets + IGPUPipelineLayout::DESCRIPTOR_SET_COUNT, nullptr);
+        return true;
+    }
+    
+    uint32_t compatibilityLimit = IGPUPipelineLayout::DESCRIPTOR_SET_COUNT;
+    if (ctx->currentState.pipeline)
+    {
+        const IGPUPipelineLayout* nextLayout = _gpipeline->getLayout();
+        compatibilityLimit = ctx->currentState.pipeline->getLayout()->isCompatibleForSet(IGPUPipelineLayout::DESCRIPTOR_SET_COUNT-1u, nextLayout);
+    }
+
+    if (compatibilityLimit == IGPUPipelineLayout::DESCRIPTOR_SET_COUNT)
+        std::fill(ctx->nextState.descriptorsParams.descSets, ctx->nextState.descriptorsParams.descSets + IGPUPipelineLayout::DESCRIPTOR_SET_COUNT, nullptr);
+    else
+    {
+        for (uint32_t i = compatibilityLimit+1u; i < IGPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
+            ctx->nextState.descriptorsParams.descSets[i] = nullptr;
+    }
+
+    return true;
+}
+
+bool COpenGLDriver::bindDescriptorSets(E_PIPELINE_BIND_POINT _pipelineType, uint32_t _first, uint32_t _count, video::IGPUDescriptorSet** _descSets, uint32_t _dynOffsetCount, const uint32_t* _dynOffsets)
+{
+    //TODO take into account `_pipelineType` param and adjust this function when compute pipelines come in
+    SAuxContext* ctx = getThreadContext_helper(false);
+    if (!ctx)
+        return false;
+
+    if (_first + _count >= IGPUPipelineLayout::DESCRIPTOR_SET_COUNT)
+        return false;
+
+    for (uint32_t i = 0u; i < _count; ++i)
+        ctx->nextState.descriptorsParams.descSets[_first+i] = core::smart_refctd_ptr<IGPUDescriptorSet>(_descSets[i]);
+
+    //TODO dynamic offsets
```

Comment:
maybe make `uint32_t _dynOffsetCount, const uint32_t* _dynOffsets` into just a smart refctd array passed by r-value, then add a smart refcounted dynamic array member `dynamicOffsets[4]` to `nextState`

This shouldn't ever be a problem as if layout changes (is incompatible) then the dynamic offsets will be dropped, its 100% the case that if the dynamic offset arrays are different lengths the pipelines will be incompatible.

you can still compare the contents of two identical length dynamic offset arrays, and decide whether to refresh the bindings or not.

OR

you can just keep a copy of the offset array (base offset + dynamic) for multibind of buffers (you'll recompute every `bindDescriptorSets`) but only ever fire off the multibind if something changes

Reply:
resolved